aws-cpp-sdk-core/source/auth/AWSCrtCredentialsProviderAdapter.cpp

```diff
@@ -0,0 +1,33 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/core/auth/AWSCrtCredentialsProviderAdapter.h>
+
+using namespace Aws::Auth;
+
+static const char* ALLOCATION_TAG = "AWSCrtCredentialsProviderAdapter";
+
+AWSCrtCredentialsProviderAdapter::AWSCrtCredentialsProviderAdapter(std::shared_ptr<Aws::Auth::AWSCredentialsProvider> credentialsProvider): credentialsProvider(std::move(credentialsProvider))
+{
+}
+
+std::shared_ptr<Aws::Crt::Auth::ICredentialsProvider> AWSCrtCredentialsProviderAdapter::GetCrtCredentialProvider() {
+    std::function<AWSCredentials()> credFunc =
+        std::bind([](const std::shared_ptr<AWSCredentialsProvider>& provider){
+            return provider->GetAWSCredentials();
+        }, this->credentialsProvider);
+
+    std::function<std::shared_ptr<Aws::Crt::Auth::Credentials>()> handler =
+        std::bind([](const std::function<AWSCredentials()>& credFunc) {
+            AWSCredentials credentials = credFunc();
+            return Aws::MakeShared<Aws::Crt::Auth::Credentials>(ALLOCATION_TAG,
+                Aws::Crt::ByteCursorFromCString(credentials.GetAWSAccessKeyId().c_str()),
```

Comment:
here ByteCursor will contain a pointer to a local object std::string c_str buffer.

Reply:
but then in c-tor of Aws::Crt::Auth::Credentials, in the function aws_credentials_new,
byte buffers will be copied before local credentials object is destroyed.

Reply:
as discussed offline, moving all of this too a anon function in the client class itself to address ownership issues, will push up once i have updated the templates